contracts/src/Summa.sol

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.18;
+
+// Uncomment this line to use console.log
+//import "hardhat/console.sol";
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+interface Verifier {
+    function verify(
+        uint256[] calldata pubInputs,
+        bytes calldata proof
+    ) external view returns (bool);
+}
+
+contract Summa is Ownable {
+    using ECDSA for bytes32;
+
+    uint256 constant ETH_INPUTS = 1;
+    uint256 constant ERC20_INPUTS = 1;
+
+    Verifier private verifier;
+
+    event ProofOfSolvencySubmitted(bytes32 indexed exchangeId);
+
+    constructor(Verifier _verifier) {
+        verifier = _verifier;
+    }
+
+    function submitProofOfSolvency(
+        string memory exchangeId,
+        address[] memory cexETHAddresses,
+        address[] memory cexERC20Addresses,
+        uint256[] memory cexETHBalances,
+        uint256[] memory cexERC20Balances,
+        address[] memory erc20ContractAddresses,
+        bytes[] memory cexEthAddressSignatures,
+        bytes[] memory cexERC20AddressSignatures,
+        uint256 mstRoot,
```

Comment:
Could you implement a feature to simplify the retrieval of `mstRoot`? 
I believe it would be very convenient to have a mapping like this: `mapping(bytes32 => uint256) public exchangeIdToMstRoot`. 
This would greatly contribute to ensuring reliability during user-side verification.

Reply:
Sure, will do!

Reply:
I gave it some thought and ended up including the MST root in the emitted event. Normally, you would want to write to the contract state only in the following cases:

- the state needs to be accessible by the contract for its future computations (e.g., in a DeFi protocol you need to store the modified state in the contract, but in our case we don't perform any computations on previous data);
- the state needs to be accessible by other contracts (again, think of DeFi composability, and we don't need any composability yet).

Events are very convenient for us because the user would probably want to verify a proof at a certain point in time that corresponds to a certain block timestamp, which is exactly how you query the contract events using filters:

```
 const filter = summaContract.filters.ProofOfSolvencySubmitted(exchangeId);
 const fetchEvents = async () => {
     const events = await summaContract.queryFilter(filter, startingBlock, endingBlock);
     ...
 };
```
It's a common practice to offload the data that's already recorded on chain (i.e., you can always find the TX and read the mstRoot function argument value) in a transaction log to save on gas.

Reply:
Yes! It seems like a better way to get info without additional costs! Thanks